File: 1151/minimum-swaps-to-group-all-1s-together.java
class Solution {
    public int minSwaps(int[] data) {
        //因為Related Topics 寫 sliding window, 就將它設成1的個數
        int ones=0;
        for(int i=0; i<data.length; i++){
            if(data[i]==1) ones++; //一邊數1
            if(i>0) data[i] += data[i-1];//一邊做 prefix sum
        }
//System.out.println(ones);
        if(ones==0) return 0;//為了 [0,0,0] 全部都是0的狀況
        int ans = ones - data[ones-1];//如果ones是0的話，不能有這行
        for(int i=ones; i<data.length; i++){
            int temp = data[i]-data[i-ones];//目前範圍內1的數量
            if(ones-temp<ans) ans = ones-temp;
        }
        return ans;
    }//出錯的原因，可能是第0筆的資料沒算到, 所以修改第10行
}//case 35/36: [1,1,1,1,0,1,0,1,1,1,1,0,1,0,0,1,1,0,0,0,1,0,0,1,1,1,0,0,1,0,0,0,1,1,1,1,1,1,0,1,0,0,1,1,1,1,1,0,1,0,1,0,1,0,1,0,1,1,1,0,1,1,1,0,0,0,0,0,1,1,1,0,0,1,0,1,0,1,1,1,0,1,1,1,0,0,0,1,1,0,1,1,1,0,1,0,1,0,0,1,0,1,1,0,0,1,0,0,1,0,1,1,0,0,1,0,0,0,0,0,1,1,1,1,0,0,1,0,0,0,0,0,0,1,0,0,1,0,1,0,1,1,1,0,0,0,1,1,1,1,1,0,1,0,0,1,1,1,1,1,1,1,1,1,0,1,1,1,0,0,1,0,1,1,1,0,0,0,0,1,0,0,1,0,0,0,0,0,1,1,0,1,0,0,1,0,1,0,1,0]

File: 1151/minimum-swaps-to-group-all-1s-together.py
# LeetCode 1151. Minimum Swaps to Group All 1's Together
# 要把 data 裡的 1 交換到「放到一起」，要「交換」幾次？
class Solution:
    def minSwaps(self, data: List[int]) -> int:
        counter = Counter(data)
        ones = counter[1]  # 總共有幾個 1，對應「最後有幾個1」需要湊在一起
        zeros = 0
        for i in range(ones):  # 先讀入前幾個數，看裡面有0「需要swap幾個外面的1進來」
            if data[i]==0: zeros += 1
        ans = zeros  # 一開始的答案「有幾個0」對應「需要swap幾次」
        for i in range(ones, len(data)):  # 利用「毛毛蟲」技巧，了解「裡面有幾個0」
            if data[i]==0: zeros += 1  # 右邊「吃到」zero
            if data[i-ones]==0: zeros -= 1  # 左邊「吐出」zero
            ans = min(ans, zeros)  # 「有幾個0」對應「需要swap幾次」，越少越好
        return ans
